kernel/src/hil/pwm.rs

```diff
@@ -0,0 +1,21 @@
+//! Interfaces for Pulse Width Modulation output.
+
+use returncode::ReturnCode;
+
+/// PWM control for a single pin.
+pub trait Pwm {
+    /// The chip-dependent type of a PWM pin.
+    type Pin;
+
+    /// Generate a PWM single on the given pin at the given frequency and duty
+    /// cycle.
+    ///
+    /// - `frequency_hz` is specified in Hertz.
+    /// - `duty_cycle` is specified in hundredths of a percent. So setting
+    ///   `duty_cycle` to 1 would mean a PWM output with a 0.01% duty cycle.
+    ///   Passing 5000 would result in a 50% duty cycle.
+    fn start(&self, pin: &Self::Pin, frequency_hz: usize, duty_cycle: usize) -> ReturnCode;
```

Comment:
What is frequency? Is it the frequency of the underlying PWM square wave, or the frequency of the sample? I'm assuming the latter, otherwise you can't effectively create analog outputs.

It would be good to specify the minimum and maximum frequencies allowed.

If one wants to construct an analog wave out of PWM outputs, how does this work? Or would you do this with a more advanced interface that allows pushing samples into a FIFO?

Do consecutive calls to start succeed, or does one have to call stop before calling start again?

I'd recommend either having the units of duty_cycle be 10^3 or 10^6; 10^4 is a weird number. If you're committed to 10^4 then I'd recommend making it a u16 rather than usize.

Reply:
> What is frequency? Is it the frequency of the underlying PWM square wave

Yes

> It would be good to specify the minimum and maximum frequencies allowed.

Isn't that hardware dependent? How does the HIL know what the max is?

> If one wants to construct an analog wave out of PWM outputs, how does this work?

They would need to add a more advanced interface.

> Do consecutive calls to start succeed, or does one have to call stop before calling start again?

Good question. Calling `start` again seems like a reasonable way to change the frequency or duty cycle.

> I'd recommend either having the units of duty_cycle be 10^3 or 10^6

Tenths of a percent didn't seem like enough granularity, hence hundredths of a percent. But that is pretty arbitrary and I don't feel strongly about it.

Reply:
"Isn't that hardware dependent? How does the HIL know what the max is?"

It could have function calls so software can determine it. It makes sense for start to fail if the specified frequency is outside the allowed range, but allowing software to determine what that range is would be useful. I've certainly had enough cases when the range specified was outside the hardware's widths and getting to the bottom of it is a pain.

The granularity can be limited by the hardware (e.g., up to 4096 slots); should querying for this also be in the HIL?